client/lib/load-script/callback-handler.js

```diff
@@ -0,0 +1,90 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+const debug = debugFactory( 'lib/load-script/callback-handler' );
+
+/**
+ * Module variables
+ */
+const callbacksForURLsInProgress = new Map();
+
+export function getCallbackMap() {
+	return callbacksForURLsInProgress;
+}
+
+export function isLoading( url ) {
+	return callbacksForURLsInProgress.has( url );
+}
+
+export function addScriptCallback( url, callback ) {
+	if ( isLoading( url ) ) {
+		debug( `Adding a callback for an existing script from "${ url }"` );
+		callbacksForURLsInProgress.get( url ).push( callback );
+	} else {
+		debug( `Adding a callback for a new script from "${ url }"` );
+		callbacksForURLsInProgress.set( url, [ callback ] );
+	}
+}
+
+export function removeScriptCallback( url, callback ) {
+	debug( `Removing a known callback for a script from "${ url }"` );
+
+	if ( ! callbacksForURLsInProgress.has( url ) ) {
+		return;
+	}
+
+	const index = callbacksForURLsInProgress.get( url ).indexOf( callback );
+
+	if ( -1 === index ) {
+		return;
+	}
+
+	if ( 1 === callbacksForURLsInProgress.get( url ).length ) {
+		callbacksForURLsInProgress.delete( url );
+		return;
+	}
+
+	callbacksForURLsInProgress.get( url ).splice( index, 1 );
+}
+
+export function removeScriptCallbacks( url ) {
+	debug( `Removing all callbacks for a script from "${ url }"` );
+	callbacksForURLsInProgress.delete( url );
+}
+
+export function removeAllScriptCallbacks() {
+	debug( 'Removing all callbacks for scripts from all URLs' );
+
+	callbacksForURLsInProgress.clear();
+}
+
+export function executeCallbacks( url, callbackArguments = null ) {
+	if ( callbacksForURLsInProgress.has( url ) ) {
+		const debugMessage = `Executing callbacks for "${ url }"`;
+		debug(
+			callbackArguments === null
+				? debugMessage
+				: debugMessage + ` with args "${ callbackArguments }"`
+		);
+
+		callbacksForURLsInProgress
+			.get( url )
+			.filter( cb => typeof cb === 'function' )
+			.forEach( cb => cb( callbackArguments ) );
+		callbacksForURLsInProgress.delete( url );
+	}
+}
+
+export function handleRequestSuccess( url ) {
```

Comment:
Maybe I'm missing something, but...

This is used as `script.onload = handleRequestSuccess`, the argument that is passed is an Event object, not a url, so the current signature is invalid IMO.

Sorry if I missed it.

Reply:
Ah, you're right - I overlooked this. I'll change the function parameters and extract the `url` from `event.target.src` instead.

Reply:
DOM access is slow, it might be better to create a map of `script` obj to `url` and lookup with `event.target` (or `this`)

Reply:
Ah, now I finally get what you were talking about from your previous comments. Investigating

Reply:
hopefully here too we shouldn't need to worry about DOM access speed much - how many script tags are we loading anyway? even if it took 5ms, which it definitely doesn't, how many total ms would that delay our customers, how many frames of UI would that set back? probably fewer than 50ms and no frames…

not to say we shouldn't optimize, but to think first about how to write maintainable code first and then optimize when performance is actually a problem that needs to be solved

Reply:
I tried a rough implementation of this and found that I could introduce it with just a few lines of code.

However, it made the code much harder to reason about. Suddenly, the `callback-handler` needed to import functions from `dom-operations` and consequently needed to mess with the internal workings of `dom-operations` to test `callback-handler` functions correctly. 

I think it'd be best to punt this idea for now; The most we're expected to load is around 10 ad scripts.